Iv 1472 powershell scripts

Maimer opened the pull request.

Commit:
IV-1472 Initial scripts started

Commit:
IV-1472 Add updates related scripts and shutdown reason

Commit:
IV-1472 Update to dev chef version number

Commit:
IV-1472 Fix metadata

Commit:
IV-1472 Add RS_TSS variable to metadata

Commit:
IV-1472 Add check to make SSC script idempotent

Commit:
IV-1472 Add setup automatic upgrade script

Commit:
IV-1472 Add upgrade script

Commit:
IV-1472 Create temporary upgrade folder to avoid conflicts

Commit:
IV-1472 Change job to process so it isn't killed with parent script

Commit:
IV-1472 Wait job test

Commit:
IV-1472 Make the upgrade function execute as its own process

Commit:
IV-1472 Updated test-script and other general formatting

Commit:
IV-1472 Change upgrade script to use SCHTASKS instead of ScheduleJob

Commit:
IV-1472 Update shutdown reason, remove scheduled job output

Commit:
IV-1472 Change version to 10.1.2

douglaswth commented on rlw/metadata.rb:
```diff
@@ -0,0 +1,129 @@
+  :recipes => ["rlw::setup_hostname", "rlw::install_updates", "install_updates_by_kb"]
+
+attribute   "WINDOWS_AUTOMATIC_UPDATES_POLICY",
+  :display_name => "SSC version to use",
```
Is this the correct display name for this input?

douglaswth commented on rlw/metadata.rb:
```diff
@@ -0,0 +1,129 @@
+  :recipes => ["rlw::install_updates_by_kb"]
+
+attribute   "SSC_SERV_VERSION",
+  :display_name => "SSC version to use",
```
This should be `SSC Serv` instead of just `SSC`.

douglaswth commented on rlw/metadata.rb:
```diff
@@ -0,0 +1,129 @@
+  :recipes => ["rlw::install_updates"]
+
+attribute   "SSC_SERV_PLATFORM",
+  :display_name => "SSC platform to use",
```
Same as above.

douglaswth commented on rlw/metadata.rb:
```diff
@@ -0,0 +1,129 @@
+  :required => "optional",
+  :type => "string",
+  :default => "3.5.0",
+  :recipes => ["rlw::install_updates"]
```
Is this the right set of recipes for this input?

douglaswth commented on rlw/metadata.rb:
```diff
@@ -0,0 +1,129 @@
+recipe      "rlw::setup_automatic_upgrade", "Periodically checks if an upgrade is available and upgrade if there is."
+recipe      "rlw::setup_hostname", "Changes the hostname of the server"
+recipe      "rlw::shutdown-reason", "Print out the reason for shutdown"
+recipe      "rlw::ssc", "Installs and configures SSC for RightScale monitoring"
```
Let's rename this recipe to `rlw::ssc-serv`.

douglaswth commented on rlw/setup_automatic_upgrade.ps1:
```diff
@@ -0,0 +1,46 @@
+$scheduledJob = SCHTASKS.exe /Query /TN 'rightlink_check_upgrade' 2> $null
+if ($env:ENABLE_AUTO_UPGRADE -eq 'false') {
+  if ($scheduledJob) {
+    SCHTASKS.exe /Delete /TN 'rightlink_check_upgrade' /F
```
~~Will this work correctly if the task is not already defined?~~

douglaswth replied:
Oh, nevermind, I see that there is an `if`/`else` around it.

douglaswth commented on rlw/setup_automatic_upgrade.ps1:
```diff
@@ -0,0 +1,46 @@
+  # Put a 0 on the front of the minute for the SCHTASKS time format
+  if ($jobMinute -lt 10) {
+    $jobMinute = "0" + $jobMinute
+  }
```
This can be simplified using PowerShell's [format operator](https://technet.microsoft.com/en-us/library/hh847732%28v=wps.620%29.aspx) `-f`:

``` powershell
$jobHour = Get-Random -Min 0 -Max 24
$jobMinute = Get-Random -Min 0 -Max 60
$jobStartTime = '{0:d2}:{1:d2}' -f $jobHour, $jobMinute
```

douglaswth commented on rlw/ssc.ps1:
```diff
@@ -0,0 +1,41 @@
+  New-ItemProperty $SSCServRegProxy Username
+  New-ItemProperty $SSCServRegProxy Password
+  New-ItemProperty $SSCServRegProxy StoreRates -Value true
+}
```
I think we need to make sure that these values are being set correctly even if they are already populated.

douglaswth commented on rlw/upgrade.ps1:
```diff
@@ -0,0 +1,167 @@
+      Exit 1
+    }
+
+    # Report to audit entry that RightLink ugpraded.
```
Typo: `ugpraded` should be `upgraded`.

douglaswth commented on rlw/upgrade.ps1:
```diff
@@ -0,0 +1,167 @@
+# Fetch information about what we should become. The "upgrades" file consists of lines formatted
+# as "currentVersion:desiredVersion". If the "upgrades" file does not exist,
+# or if the current version is not in the file, no upgrade is done.
+$upgradesFile = "${PSScriptRoot}\upgrades"
```
Do we really need to save this in a file? Would it be okay to just do `DownloadString` instead?

douglaswth commented on rlw/upgrade.ps1:
```diff
@@ -0,0 +1,167 @@
+if (!(Test-Path -Path $TMP_DIR)) {
+  New-Item -Path $TMP_DIR -Type Directory | Out-Null
+}
+
```
Are the following few conditional `Remove-Item`s to clear out the temp directory before putting stuff in it? If so, there should probably be a comment here explaining that.

Maimer replied:
Comment was added, not sure why it isn't updating this as a comment on an outdated diff.

douglaswth commented on rlw/upgrade.ps1:
```diff
@@ -0,0 +1,167 @@
+  # Fork a new task since this main process is started by RightLink and we are restarting it.
+  Start-Process Powershell -ArgumentList "-Command & { $upgradeFunction upgradeRightLink ${currentVersion} ${desiredVersion} }"
+} else {
+  Write-Output "Updated version does not appear to be desired version: ${newVersion}"
```
Shouldn't this exit with an error code?

Maimer replied:
Exit code was added, not sure why it isn't updating this as a comment on an outdated diff.

douglaswth commented on rlw/wait-for-eip.ps1:
```diff
@@ -0,0 +1,64 @@
+# spend at most 15 minutes checking the API hosts for either the expected IP address or an incorrect IP address
+$startTime = Get-Date
+$wc = New-Object System.Net.WebClient
+while ((New-TimeSpan -Start $startTime -End (Get-Date)).TotalSeconds -lt 900) {
```
Why not use `.TotalMinutes -lt 15`?

Commit:
IV-1472 Iteration on Review

Fix up errors in metadata
Renamed ssc.ps1 to ssc-serv.ps1
Changed formatting for automatic upgrade time
Make SSC Serv script work properly if rerun
Upgrade now downloads string instead of file
Various Minor Fixes to comments and other code

douglaswth commented on rlw/setup_automatic_upgrade.ps1:
```diff
@@ -0,0 +1,39 @@
+    Write-Output 'Recreating schedule job'
+    SCHTASKS.exe /Change /RU 'SYSTEM' /TN 'rightlink_check_upgrade' /ST $jobStartTime
+  } else {
+    SCHTASKS.exe /Create /RU 'SYSTEM' /ST ${jobHour}:${jobMinute} /SC DAILY `
```
Shouldn't `${jobHour}:${jobMinute}` actually be `$jobStartTime` like above?

douglaswth commented on rlw/ssc-serv.ps1:
```diff
@@ -0,0 +1,46 @@
+} else {
+  Set-ItemProperty $SSCServRegProxy URL -Value "http://localhost:${ProxyPort}/rll/tss/collectdv5"
+  Set-ItemProperty $SSCServRegProxy Username
+  Set-ItemProperty $SSCServRegProxy Password
```
According to [the documentation for Set-ItemProperty](https://technet.microsoft.com/en-us/library/hh849844%28v=wps.620%29.aspx) and running it in the PowerShell prompt, the value parameter is required. Did you actually test this code path?

douglaswth commented on rlw/upgrade.ps1:
```diff
@@ -0,0 +1,167 @@
+      Exit 1
+    }
+
+    # Report to audit entry that RightLink uggraded.
```
Typo: `uggraded` should be `upgraded`.

douglaswth commented on rlw/upgrade.ps1:
```diff
@@ -0,0 +1,167 @@
+# as "currentVersion:desiredVersion". If the "upgrades" file does not exist,
+# or if the current version is not in the file, no upgrade is done.
+$desiredVersion = (New-Object System.Net.WebClient).DownloadString($env:UPGRADES_FILE_LOCATION) |
+                  Select-String "\s*${currentVersion}\s*:\s*(\S+)\s*" | % { $_.matches.Groups[1].value }
```
Shouldn't there be `^`/`$` in this regexp?

Commit:
IV-1472 Further Iteration

Fix jobStartTime for creating new scheduled task
Added Value for item properties in ssc serv script
Fixed upgrade file download regex matching
